Adding an ordering to search result items. Adding styling to results …

dkelly-sighter opened the pull request:
Adding an ordering to search result items. Adding styling to results that
bold and italicizes where the search term matches on the name. Adding a
default height and width so that the search results appear full on load,
instead of leaving it to dynamic behavior.

dkelly-sighter commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchUiImpl.java:
```diff
@@ -122,8 +123,11 @@ public void accept(final List<SearchResult<?>> results) {
       scrollPane.setBorder(BorderFactory.createEmptyBorder());
       final JPopupMenu popup = new JPopupMenu();
       popup.add(scrollPane);
+      popup.setPopupSize(searchBox.getWidth(), 300);
```
Not sure how I feel about this, but being dynamic with the height seemed a bit beyond the scope of the changes.

dkelly-sighter commented on ganttproject/src/net/sourceforge/ganttproject/search/TaskSearchService.java:
```diff
@@ -42,20 +42,33 @@ public TaskSearchService() {
+      if (isNotEmptyAndContains(t.getNotes(), query)) {
+        matched = true;
+        snippet = "Notes";
+        snippetText = t.getNotes();
```
If a note is super long it is possible for the text to go on and on. Wasn't sure how to tackle that. Matching a keyword on text and having ellipses around complete words so many characters to the right and left opens up a whole ton of fringe cases.

dkelly-sighter commented:
@dbarashev ping

dbarashev commented:
Hi, thanks for reminding :)
Cool stuff. I made a quick test, here is a couple of bugs:

Task IDs (?) seem to be wrong, e.g. on the screenshot below (it is a sample project) there is Task#1 in the list, but the matching task has id=14

![image](https://user-images.githubusercontent.com/2028330/37840977-89f2ddcc-2ecf-11e8-89b3-bbf1a761c14f.png)

The same matching note attaches to all matching tasks, see screenshot

![image](https://user-images.githubusercontent.com/2028330/37841074-ccc1ab2e-2ecf-11e8-8b44-463d8a08a2ea.png)

The same note appears under all task entries

dkelly-sighter commented:
Oh dear me. Yeah, lemme fix those things.

dkelly-sighter commented:
Apologies. The Task # was a misunderstanding of your proposed requirement. I thought you wanted the number according to the result count, not the id of the actual Task/Resource. Switched. Also fixed the notes appearing multiple times.

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResult.java:
```diff
@@ -19,28 +19,51 @@
   private final String myOrigin;
-  private final String mySnippet;
-  private final String myLabel;
+  private final String secondaryLabel;
```
May I ask you to change the names according to the surrounding style? We use prefix `my` for member variables and do not use `this`.

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResult.java:
```diff
@@ -19,28 +19,51 @@
-  public SearchResult(String label, String snippet, String origin, SearchObjectType searchObject,
-      SearchService<?, SearchObjectType> service) {
+  public SearchResult(
+          int id,
```
Also style comment: we use two-space indents and four-space line continuation.

dkelly-sighter replied:
That's what I get for using netbeans instead of IntelliJ. Fixing.

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResult.java:
```diff
@@ -19,28 +19,51 @@
+          String myQueryMatch,
+          String secondaryLabel,
+          String secondaryText,
+          String origin,
```
My IntelliJ tells me that origin is not used. Remove? Also, visibility scope of constructor nearly all getters can be changed to package-private

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResult.java:
```diff
@@ -19,28 +19,51 @@
+
+  public String getSecondaryText() { return secondaryText; }
+
+  public String getMyQueryMatch() { return myQueryMatch; }
```
One more style comment: in getters we remove `my`. This `my` prefix is actually a replacement of `this.`, not a part of the name semantics.

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResultCellRenderer.java:
```diff
@@ -0,0 +1,47 @@
+package net.sourceforge.ganttproject.search;
```
Please add a copyright notice:
```
/*
Copyright 2018 BarD Software s.r.o, Douglas Kelly

This file is part of GanttProject, an opensource project management tool.
GanttProject is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
GanttProject is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with GanttProject.  If not, see <http://www.gnu.org/licenses/>.
*/
```

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResultCellRenderer.java:
```diff
@@ -0,0 +1,47 @@
+
+public class SearchResultCellRenderer implements ListCellRenderer {
+
+    protected DefaultListCellRenderer defaultRenderer = new DefaultListCellRenderer();
```
This can be private I think

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResultCellRenderer.java:
```diff
@@ -0,0 +1,47 @@
+
+        if (o instanceof SearchResult) {
+            SearchResult searchResult = (SearchResult) o;
+            theText.append("<b>" + searchResult.getTypeOfResult()+ " #" + searchResult.getId() + "</b>: ");
```
This should be internationalized, although admittedly it is not the easiest thing.

In general, we do i18n like this:
`GanttLanguage.getInstance().formatText(key, arg1, arg2)`. The implementation searches for the key value in appropriate resource bundle and [feeds the value and args](https://github.com/bardsoftware/ganttproject/blob/3ca29c5ba6facaf4b1fb8414bb831a8fcec0a096/ganttproject/src/net/sourceforge/ganttproject/language/GanttLanguage.java#L404) to `MessageFormat.format()`

I suggest extracting HTML markup of both labels into separate key-value pairs in the i18n resources and passing result type, id and label with highlighted match as arguments.  [Existing keys related to Search function](https://github.com/bardsoftware/ganttproject/blob/3ca29c5ba6facaf4b1fb8414bb831a8fcec0a096/ganttproject/data/resources/language/i18n.properties#L563) are prefixed with `search.` so please use the same prefix:

```
search.prompt.shortcut = <Ctrl+F>
search.result.line1 = <b>{0} #{1}</b>: {2}<br>
search.result.line2 = <b>{0}</b>: {1}
```

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResultCellRenderer.java:
```diff
@@ -0,0 +1,47 @@
+        }
+        theText.append("</html>");
+        renderer.setText(theText.toString());
+        renderer.setFont(new Font("Helvetica", Font.PLAIN, 12));
```
Is this font change required? If possible, please leave the default font because we should respect user font size settings, DPI and language (think Chinese or Japanese glyphs which are unlikely to be in Helvetica)

dkelly-sighter replied:
Roger, no need for it.

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/TaskSearchService.java:
```diff
@@ -29,8 +29,8 @@
-    public MySearchResult(Task t, TaskSearchService searchService) {
-      super("Task: " + t.getName(), "", "", t, searchService);
+    public MySearchResult(Task t, TaskSearchService searchService, String query, String snippet, String snippetText) {
+      super(t.getTaskID(), "Task", t.getName(), query, snippet, snippetText, "", t, searchService);
```
Admittedly, this was not i18n'ed before the changes, but while you're on it, can you i18n these "Task", "Resource" and "Notes" labels? We already have keys in 18n.properties fpr all of them, so it basically comes to just calling `GanttLanguage.getInstance().getText()`

dkelly-sighter replied:
I'm seeing the plurals for Tasks and Resources, but not the singulars. I'll add those in.

dkelly-sighter replied:
Oh, I guess it isn't that simple? I'll go plural for now.

dkelly-sighter commented:
Sweet. I'll get er done.

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/TaskSearchService.java:
```diff
@@ -24,13 +24,14 @@
-    public MySearchResult(Task t, TaskSearchService searchService, String query, String snippet, String snippetText) {
-      super(t.getTaskID(), "Task", t.getName(), query, snippet, snippetText, "", t, searchService);
+    MySearchResult(Task t, TaskSearchService searchService, String query, String snippet, String snippetText) {
+      super(t.getTaskID(), GanttLanguage.getInstance().getCorrectedLabel("task"), t.getName(), query, snippet, snippetText, t, searchService);
```
Apparently I was wrong about existing keys for resources and tasks, sorry. They do exist in plural form but not in singular. So let's create new keys: `generic.task` and `generic.resource`

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResultCellRenderer.java:
```diff
@@ -8,13 +26,9 @@
+    private DefaultListCellRenderer defaultRenderer = new DefaultListCellRenderer();
 
     @Override
     public Component getListCellRendererComponent(JList jList, Object o, int index, boolean b, boolean b1) {
```
Please give mnemonic names to these args (isSelected, cellHasFocus...)

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResultCellRenderer.java:
```diff
@@ -0,0 +1,60 @@
+            theText.append(o.toString());
+        }
+        theText.append("</html>");
+        renderer.setText(theText.toString());
```
UI tip: add `renderer.setBorder(BorderFactory.createEmptyBorder(5, 3, 5, 3));` here

and in `SearchUiImpl.PopupSearchCallback` replace the type of `list`: JList => JXList 
and add `list.setHighlighters(UIUtil.ZEBRA_HIGHLIGHTER);`

Altogether this will add some padding in the list cells and zebra-highlighting of the cells, so that it will be easier to distinguish between neighbor cells

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResultCellRenderer.java:
```diff
@@ -0,0 +1,60 @@
+import javax.swing.ListCellRenderer;
+
+public class SearchResultCellRenderer implements ListCellRenderer {
+
```
Please change indent to 2 spaces.

dkelly-sighter commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResult.java:
```diff
@@ -19,55 +19,48 @@
-          SearchObjectType searchObject,
-          SearchService<?, SearchObjectType> service
+  SearchResult(
+    int id,
```
K, I think I've got this correct now after being asked twice. I guess we'll see with the next commit.

dbarashev commented on ganttproject/data/resources/language/i18n.properties:
```diff
@@ -163,6 +163,8 @@ ganttReport = GanttProject Report
 general = General
 generic.endDate.label = End date
 generic.startDate.label = Start date
+generic.resource = Resource
```
final: one line up please :) So that it was ordered lexicographically

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/SearchResultCellRenderer.java:
```diff
@@ -0,0 +1,58 @@
+package net.sourceforge.ganttproject.search;
+
+/*
```
Please move it to the very beginning of the file, before `package` statement.

dbarashev commented on ganttproject/src/net/sourceforge/ganttproject/search/TaskSearchService.java:
```diff
@@ -41,21 +42,34 @@ public TaskSearchService() {
+      }
+      if (isNotEmptyAndContains(String.valueOf(t.getTaskID()), query)) {
         matched = true;
+        snippet = "Id";
```
Please i18n this also. "id" must be a good key
